README
Xtables-addons
==============

Xtables-addons is the proclaimed successor to patch-o-matic(-ng). It
contains extensions that were not accepted in the main Xtables
package.

Xtables-addons is different from patch-o-matic in that you do not
have to patch or recompile either kernel or Xtables(iptables). But
please see the INSTALL file for the minimum requirements of this
package.

All code imported from patch-o-matic has been reviewed and all
apparent bugs like binary stability across multiarches, missing
sanity checks and incorrect endianess handling have been fixed,
simplified, and sped up.


Included in this package
========================
- ipset 6.10-genl
- xt_ACCOUNT 1.16, libxt_ACCOUNT 1.3


Inclusion into a kernel tree
============================




External extensions
===================

The program "xa-download-more" can be used to download more
extensions from 3rd parties into the source tree. The URLs are listed
in the "sources" file. If the "sources" file contains an entry like

	http://foobar.org/xa/

xa-download-more will inspect http://foobar.org/xa/xa-index.txt for
files to download. That file may contain

	foobar.tar.bz2

and xa-download-more will then retrieve and unpack
http://foobar.org/xa/foobar.tar.bz2.

Files that should be contained in the tarball are an mconfig and
Kbuild files to control building the extension, libxt_foobar.c for
the userspace extension and xt_foobar.c for the kernel extension.

	mconfig.foobar
	extensions/Kbuild.foobar
	extensions/Mbuild.foobar
	extensions/libxt_foobar.c
	extensions/libxt_foobar.man
	extensions/xt_foobar.c
	extensions/xt_foobar.h


doc/changelog.txt

HEAD
====
Fixes:
- compat_xtables: fixed mistranslation of checkentry return values
  (affected kernels < 2.6.23)


v1.41 (2012-01-04)
==================
Changes:
- Deactivate build of ipset-genl by default.
  I think the original ipset package can now take over, given there are
  a handful of kernels (2.6.39 onwards) that do not need patching.
Enhancements:
- Support for Linux 3.2


v1.40 (2011-11-30)
==================
Fixes:
- build: the code actually requires at least iptables 1.4.5 (would yield a
  compile error otherwise), make sure configure checks for it; update INSTALL
- xt_ECHO: fix kernel warning about RTAX_HOPLIMIT being used
- xt_ipv4options: fix an infinite loop
Changes:
- xt_ECHO: now calculates UDP checksum
- Linux kernel versions below 2.6.32 are no longer officially
  supported, and will not be part of compilation testing.
- update to ipset 6.10
Enhancements:
- xt_ECHO: IPv6 support


v1.39 (2011-09-21)
==================
Fixes:
- libxt_ACCOUNT: fix compilation after missing libxtables_CFLAGS
- build: fix compilation after missing libxtables_CFLAGS in submodules
- build: add missing linux/version.h includes where needed
Changes:
- Remove unsupported ipset 4.x from the Xtables-addons distribution
- ipset: move ipset_errcode from src to library to avoid undefined reference
- update to ipset 6.9.1


v1.38 (2011-08-20)
==================
- xt_CHECKSUM: abort build when the feature is already provided by mainline
- xt_SYSRQ: fix UDPLITE header lookup in IPv6
- xt_TARPIT: fix kernel warning about RTAX_HOPLIMIT being used
- xt_TEE: abort build when the feature is already provided by mainline
- xt_ipp2p: support UDPLITE
- xt_pknock: support UDPLITE
- xt_psd: restore functionality with UDP
- xt_psd: support UDPLITE
- update to ipset 6.8
- support for Linux 3.1


v1.37 (2011-06-25)
==================
Fixes:
- xt_SYSRQ: make IPv6 trigger work again
- xt_SYSRQ: improve security: include host address in digest
- xt_TARPIT: fix a kernel oops in --reset mode


v1.36 (2011-06-03)
==================
Changes:
- xt_geoip: avoid recursive function calls
- xt_TARPIT: unlock for use in all tables
- xt_TARPIT: honeypot and reset modes
- update to ipset 6.7
- support for Linux 3.0


v1.35 (2011-04-11)
==================
Enhancements:
- update to ipset 6.3
  * allow "new" as a commad alias to "create"
  * resolving IP addresses did not work at listing/saving sets, fixed
  * check ICMP and ICMPv6 with the set match and target in the testsuite
  * avoid possible syntax clashing at saving hostnames
  * fix linking with CONFIG_IPV6=n
  * sctp, udplite support for the hash:*port* types
- ipset-genl: handle EAGAIN return value emitted from autoloader
- ipset-genl: resolve nfgenmsg remains and fix spurious protocol abort


v1.34 (2011-04-07)
==================
Fixes:
- xt_pknock: avoid crash when hash TFM could not be allocated
- xt_pknock: avoid inversion of rule lookup that led to warnings
- xt_DNETMAP: add missing module alias
- xt_DNETMAP: support for kernels below 2.6.34
Changes:
- Linux kernel versions below 2.6.29 are no longer officially
  supported, and will not be part of compilation testing.


v1.33 (2011-02-02)
==================
Fixes:
- build: restore functionality of `configure --without-kbuild`
- build: fix objdir builds for ipset-5 (xt-a specific)
- build: fix missing inclusion of dependency rules
- xt_LOGMARK: fix detection of untracked connection for Linux >= 2.6.36
Enhancements:
- IPv6 support for xt_geoip
- Update to ipset 5.3
  * make IPv4 and IPv6 address handling similar
  * show correct line numbers in restore output for parser errors
- Update to ipset 5.4
  * fixed ICMP and ICMPv6 handling
  * fixed trailing whitespaces and pr_* messages
  * fixed module loading at create/header commands
- build: support for Linux up to 2.6.38
- build: preliminary support for iptables 1.4.11


v1.32 (2011-01-04)
==================
Fixes:
- Update to ipset 4.5
  * the iptreemap type used wrong gfp flags when deleting entries
- Include ipset 5.2 with genetlink patch (beta)
  * no kernel patch needed, but requires Linux >= 2.6.35
    and thus needs to be manually enabled in mconfig


v1.31 (2010-11-05)
==================
Fixes:
- build: improve detection of kernel version and error handling
Changes:
- build: automatically derive Xtables module directory, thus
  --with-xtlibdir is no longer needed for ./configure in most cases
  (If I still see a distro using it, I will scold you for not
  reading this changelog.)
Enhancements:
- LOGMARK: print remaining lifetime of cts
- xt_iface: allow matching against incoming/outgoing interface
- libxt_gradm: match packets based on status of grsecurity RBAC
  (userspace part only - xt_gradm is in the grsec patch)


v1.30 (2010-010-02)
===================
Fixes:
- update to ipset 4.4
  * ipport{,ip,net}hash did not work with mixed "src" and "dst"
    destination parameters
Changes:
- deactivate building xt_TEE and xt_CHECKSUM by default, as these have been
  merged upstream in Linux 2.6.35 and 2.6.36, respectively.
  Distros still wishing to build this need to enable it in their build
  script, e.g. perl -i -pe 's{^build_TEE=.*}{build_TEE=m}' mconfig;


v1.29 (2010-09-29)
==================
- compat_xtables: return bool for match_check and target_check in 2.6.23..34
- ipset: enable building of ip_set_ipport{ip,net}hash.ko
- support for Linux 2.6.36
- SYSRQ: resolve compile error with Linux 2.6.36
- TEE: resolve compile error with Linux 2.6.36
- add workaround for broken linux-glibc-devel 2.6.34 userspace headers
  ("implicit declaration of function 'ALIGN'")


v1.28 (2010-07-24)
==================
- RAWNAT: IPv6 variants erroneously rejected masks /33-/128
- new target xt_CHECKSUM
- xt_length2: add support for IPv6 jumbograms
- xt_geoip: fix possible out-of-bounds access
- import xt_geoip database scripts


v1.27 (2010-05-16)
==================
- further updates for the upcoming 2.6.35 changes


v1.26 (2010-04-30)
==================

extensions/xt_psd.c
/*
  This is a module which is used for PSD (portscan detection)
  Derived from scanlogd v2.1 written by Solar Designer <solar@false.com>
  and LOG target module.

  Copyright (C) 2000,2001 astaro AG

  This file is distributed under the terms of the GNU General Public
  License (GPL). Copies of the GPL can be obtained from:
     ftp://prep.ai.mit.edu/pub/gnu/GPL

  2000-05-04 Markus Hennig <hennig@astaro.de> : initial
  2000-08-18 Dennis Koslowski <koslowski@astaro.de> : first release
  2000-12-01 Dennis Koslowski <koslowski@astaro.de> : UDP scans detection added
  2001-01-02 Dennis Koslowski <koslowski@astaro.de> : output modified
  2001-02-04 Jan Rekorajski <baggins@pld.org.pl> : converted from target to match
  2004-05-05 Martijn Lievaart <m@rtij.nl> : ported to 2.6
  2007-04-05 Mohd Nawawi Mohamad Jamili <nawawi@tracenetworkcorporation.com> : ported to 2.6.18
  2008-03-21 Mohd Nawawi Mohamad Jamili <nawawi@tracenetworkcorporation.com> : ported to 2.6.24
  2009-08-07 Mohd Nawawi Mohamad Jamili <nawawi@tracenetworkcorporation.com> : ported to xtables-addons
*/

#define pr_fmt(x) KBUILD_MODNAME ": " x
#include <linux/module.h>
#include <linux/moduleparam.h>
#include <linux/skbuff.h>
#include <linux/ip.h>
#include <net/tcp.h>
#include <linux/spinlock.h>
#include <linux/netfilter_ipv4/ip_tables.h>
#include <linux/netfilter/x_tables.h>
#include "xt_psd.h"
#include "compat_xtables.h"

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Dennis Koslowski <koslowski@astaro.com>");
MODULE_AUTHOR("Martijn Lievaart <m@rtij.nl>");
MODULE_AUTHOR("Jan Rekorajski <baggins@pld.org.pl>");
MODULE_AUTHOR(" Mohd Nawawi Mohamad Jamili <nawawi@tracenetworkcorporation.com>");
MODULE_DESCRIPTION("Xtables: PSD - portscan detection");
MODULE_ALIAS("ipt_psd");

#define HF_DADDR_CHANGING   0x01
#define HF_SPORT_CHANGING   0x02
#define HF_TOS_CHANGING	    0x04
#define HF_TTL_CHANGING	    0x08

/*
 * Information we keep per each target port
 */
struct port {
	u_int16_t number;      /* port number */
	u_int8_t proto;        /* protocol number */
	u_int8_t and_flags;    /* tcp ANDed flags */
	u_int8_t or_flags;     /* tcp ORed flags */
};

/*
 * Information we keep per each source address.
 */
struct host {
	struct host *next;						/* Next entry with the same hash */
	unsigned long timestamp;					/* Last update time */
	struct in_addr src_addr;				/* Source address */
	struct in_addr dest_addr;				/* Destination address */
	unsigned short src_port;				/* Source port */
	int count;								/* Number of ports in the list */
	int weight;								/* Total weight of ports in the list */
	struct port ports[SCAN_MAX_COUNT - 1];	/* List of ports */
	unsigned char tos;						/* TOS */
	unsigned char ttl;						/* TTL */
	unsigned char flags;					/* HF_ flags bitmask */
};

/*
 * State information.
 */
static struct {
	spinlock_t lock;
	struct host list[LIST_SIZE];	/* List of source addresses */
	struct host *hash[HASH_SIZE];	/* Hash: pointers into the list */
	int index;						/* Oldest entry to be replaced */
} state;

/*
 * Convert an IP address into a hash table index.
 */
static inline int hashfunc(struct in_addr addr)
{
	unsigned int value;
	int hash;

	value = addr.s_addr;
	hash = 0;
	do {
		hash ^= value;
	} while ((value >>= HASH_LOG) != 0);

	return hash & (HASH_SIZE - 1);
}

static bool
xt_psd_match(const struct sk_buff *pskb, struct xt_action_param *match)
{
	const struct iphdr *iph;
	const struct tcphdr *tcph = NULL;
	const struct udphdr *udph;
	union {
		struct tcphdr tcph;
		struct udphdr udph;
	} _buf;
	struct in_addr addr;
	u_int16_t src_port,dest_port;
  	u_int8_t tcp_flags, proto;
	unsigned long now;
	struct host *curr, *last, **head;
	int hash, index, count;
	/* Parameters from userspace */
	const struct xt_psd_info *psdinfo = match->matchinfo;

	/* IP header */
	iph = ip_hdr(pskb);

	/* Sanity check */
	if (iph->frag_off & htons(IP_OFFSET)) {
		pr_debug("sanity check failed\n");
		return false;
	}

	/* TCP or UDP ? */
	proto = iph->protocol;
	/* Get the source address, source & destination ports, and TCP flags */

	addr.s_addr = iph->saddr;
	/* We're using IP address 0.0.0.0 for a special purpose here, so don't let
	 * them spoof us. [DHCP needs this feature - HW] */
	if (addr.s_addr == 0) {
		pr_debug("spoofed source address (0.0.0.0)\n");
		return false;
	}

	if (proto == IPPROTO_TCP) {
		tcph = skb_header_pointer(pskb, match->thoff,
		       sizeof(_buf.tcph), &_buf.tcph);
		if (tcph == NULL)
			return false;

		/* Yep, it's dirty */
		src_port = tcph->source;
		dest_port = tcph->dest;
		tcp_flags = *((u_int8_t*)tcph + 13);
	} else if (proto == IPPROTO_UDP || proto == IPPROTO_UDPLITE) {
		udph = skb_header_pointer(pskb, match->thoff,
		       sizeof(_buf.udph), &_buf.udph);
		if (udph == NULL)
			return false;
		src_port  = udph->source;
		dest_port = udph->dest;
		tcp_flags = 0;
	} else {
		pr_debug("protocol not supported\n");
		return false;
	}

	/* Use jiffies here not to depend on someone setting the time while we're
	 * running; we need to be careful with possible return value overflows. */
	now = jiffies;

	spin_lock(&state.lock);

	/* Do we know this source address already? */
	count = 0;
	last = NULL;
	if ((curr = *(head = &state.hash[hash = hashfunc(addr)])) != NULL)
		do {
			if (curr->src_addr.s_addr == addr.s_addr)
				break;
			count++;
			if (curr->next != NULL)
				last = curr;
		} while ((curr = curr->next) != NULL);

	if (curr != NULL) {

		/* We know this address, and the entry isn't too old. Update it. */
		if (now - curr->timestamp <= (psdinfo->delay_threshold*HZ)/100 &&
		    time_after_eq(now, curr->timestamp)) {

			/* Just update the appropriate list entry if we've seen this port already */
			for (index = 0; index < curr->count; index++) {
				if (curr->ports[index].number == dest_port) {
					curr->ports[index].proto = proto;
					curr->ports[index].and_flags &= tcp_flags;
					curr->ports[index].or_flags |= tcp_flags;
					goto out_no_match;
				}
			}

			/* TCP/ACK and/or TCP/RST to a new port? This could be an outgoing connection. */
			if (proto == IPPROTO_TCP && (tcph->ack || tcph->rst))
				goto out_no_match;

			/* Packet to a new port, and not TCP/ACK: update the timestamp */
			curr->timestamp = now;

			/* Logged this scan already? Then drop the packet. */
			if (curr->weight >= psdinfo->weight_threshold)
				goto out_match;

			/* Specify if destination address, source port, TOS or TTL are not fixed */
			if (curr->dest_addr.s_addr != iph->daddr)
				curr->flags |= HF_DADDR_CHANGING;
			if (curr->src_port != src_port)
				curr->flags |= HF_SPORT_CHANGING;
			if (curr->tos != iph->tos)
				curr->flags |= HF_TOS_CHANGING;
			if (curr->ttl != iph->ttl)
				curr->flags |= HF_TTL_CHANGING;

			/* Update the total weight */
			curr->weight += (ntohs(dest_port) < 1024) ?
				psdinfo->lo_ports_weight : psdinfo->hi_ports_weight;

			/* Got enough destination ports to decide that this is a scan? */
			/* Then log it and drop the packet. */
			if (curr->weight >= psdinfo->weight_threshold)
				goto out_match;

			/* Remember the new port */
			if (curr->count < SCAN_MAX_COUNT) {
				curr->ports[curr->count].number = dest_port;
				curr->ports[curr->count].proto = proto;
				curr->ports[curr->count].and_flags = tcp_flags;
				curr->ports[curr->count].or_flags = tcp_flags;
				curr->count++;
			}

			goto out_no_match;
		}

		/* We know this address, but the entry is outdated. Mark it unused, and
		 * remove from the hash table. We'll allocate a new entry instead since
		 * this one might get re-used too soon. */
		curr->src_addr.s_addr = 0;
		if (last != NULL)
			last->next = last->next->next;
		else if (*head != NULL)
			*head = (*head)->next;
		last = NULL;
	}

	/* We don't need an ACK from a new source address */
	if (proto == IPPROTO_TCP && tcph->ack)
		goto out_no_match;

	/* Got too many source addresses with the same hash value? Then remove the
	 * oldest one from the hash table, so that they can't take too much of our
	 * CPU time even with carefully chosen spoofed IP addresses. */
	if (count >= HASH_MAX && last != NULL)
		last->next = NULL;

	/* We're going to re-use the oldest list entry, so remove it from the hash
	 * table first (if it is really already in use, and isn't removed from the
	 * hash table already because of the HASH_MAX check above). */

	/* First, find it */
	if (state.list[state.index].src_addr.s_addr != 0)
		head = &state.hash[hashfunc(state.list[state.index].src_addr)];
	else
		head = &last;
	last = NULL;
	if ((curr = *head) != NULL)
		do {
			if (curr == &state.list[state.index])
				break;
			last = curr;
		} while ((curr = curr->next) != NULL);

	/* Then, remove it */
	if (curr != NULL) {
		if (last != NULL)
			last->next = last->next->next;
		else if (*head != NULL)
			*head = (*head)->next;
	}

	/* Get our list entry */
	curr = &state.list[state.index++];
	if (state.index >= LIST_SIZE)
		state.index = 0;

	/* Link it into the hash table */
	head = &state.hash[hash];
	curr->next = *head;
	*head = curr;

	/* And fill in the fields */
	curr->timestamp = now;
	curr->src_addr = addr;
	curr->dest_addr.s_addr = iph->daddr;
	curr->src_port = src_port;
	curr->count = 1;
	curr->weight = (ntohs(dest_port) < 1024) ? psdinfo->lo_ports_weight : psdinfo->hi_ports_weight;
	curr->ports[0].number = dest_port;
	curr->ports[0].proto = proto;
	curr->ports[0].and_flags = tcp_flags;
	curr->ports[0].or_flags = tcp_flags;
	curr->tos = iph->tos;
	curr->ttl = iph->ttl;

out_no_match:
	spin_unlock(&state.lock);
	return false;

out_match:
	spin_unlock(&state.lock);
	return true;
}

static struct xt_match xt_psd_reg __read_mostly = {
	.name		= "psd",
	.family    = NFPROTO_IPV4,
	.revision  = 1,
	.match		= xt_psd_match,
	.matchsize	= sizeof(struct xt_psd_info),
	.me			= THIS_MODULE,
};

static int __init xt_psd_init(void)
{
	spin_lock_init(&(state.lock));
	return xt_register_match(&xt_psd_reg);
}

static void __exit xt_psd_exit(void)
{
        xt_unregister_match(&xt_psd_reg);
}

module_init(xt_psd_init);
module_exit(xt_psd_exit);

